File: Frontend/Interviews/2021Interviews/ts/partOne/Polymorphism.ts
interface IUsb {
    transferData(): void
}

class Disk implements IUsb {
    private name: string

    constructor(name: string) {
        this.name = name
    }

    transferData(): void {
        console.log(`${this.name} is transfer`)
    }
}


class Phone implements IUsb {
    private name: string

    constructor(name: string) {
        this.name = name
    }

    transferData(): void {
        console.log(`${this.name} is transfer`)
    }
}


class Computer {
    transfer(usb: IUsb) {
        usb.transferData()
    }
}

const computer = new Computer();
const disk = new Disk("disk");
const phone = new Phone("iPhone");

computer.transfer(disk);
computer.transfer(phone);

File: Frontend/NativeApps/Dart/src/oop/Polymorphism.dart
abstract class USB {
  String deviceName;

  void transferData();
}

//手机类, 实现传输数据功能
class Phone implements USB {
  @override
  String deviceName;

  Phone(this.deviceName);

  @override
  void transferData() {
    print('${this.deviceName} is transfer');
  }
}

//相机类,实现传输数据功能
class Camera implements USB {
  @override
  String deviceName;

  Camera.name(this.deviceName);

  @override
  void transferData() {
    print('${this.deviceName} is transfer');
  }
}

class Computer {
  //电脑, 能给具备传输数据的设备进行数据传输
  void transfer(USB device) {
    device.transferData();
  }
}

void main() {
  Phone phone = new Phone("iphone");
  Camera camera = new Camera.name("sony");
  Computer computer = new Computer();

  computer.transfer(phone); //iphone is transfer
  computer.transfer(camera); //sony is transfer
}
